### DISCUSSION TITLE
Where should the .env file be placed?

### DISCUSSION THREAD

**Discussion by SquireOfSoftware** [allan2/dotenvy#60] [Q&A] [answered]
I feel like this is a fairly obvious question but I am also fairly certain that my rust/crate naivety is also at play here.

## Where should the .env file be placed?

I have a project that I have been tinkering away with and I have it set up like so:

```
/
/.env <-- has the env var `DB_USER=test`
/graphql_api
/graphql_api/.env <-- has the env var `DB_USER=api`
/front_end
/front_end/.env <-- does NOT have the env var `DB_USER`
```

And I noticed that if I run: `cargo run -p graphql_api` from the root directory it will get the env var of `DB_USER=test` from the root directory, but if I run `cargo run -p graphql_api` from the `graphql_api` directory then it will pick up `DB_USER=api`.

So I have two general questions:
1. Is it expect that the `.env` file is based on the location of where you run your cargo command?
2. Is it best practice to have workspaces that have their own `.env` files?
3. Is there a way for me to point `dotenv` under the workspace to use the "workspaces" own `.env` file? Or is the expectation that you run your cargo command when you are in the right location and therefore this question is just plain moot?

Feel free to let me know if these questions are also invalid because I have just come off the Rust book fairly recently and was curious about the actual behaviour of `dotenv`...but recently discovered that it was unmaintained and the torch has been sorta passed to this repo? 

And also feel free to point me to documentation to read since it seems my googlefu is failing me on these questions.

**Comment by allan2** [comment, marked as answer]
Hi @SquireOfSoftware,

Docs are on [docs.rs](https://docs.rs/dotenvy/latest/dotenvy/). When a crate is published to crates.io, the [docs are automatically generated](https://doc.rust-lang.org/beta/rust-by-example/meta/doc.html).


1. You can put the *.env* file(s) wherever you want. `dotenv()` picks up the **first** file named *.env*, looking in the current directory and then upward in the parent directories.
2. I'm not sure if I would call it a best practice, but I like to have a single *.env* file at the project root for simplicity.
If you are mimicking a production setup and would like to keep separate sets of environment variables, you could do:
```rust
dotenvy::from_path("/path/to/project/.env.backend")?;
dotenvy::from_path("/path/to/project/.env.frontend")?;
```

3. There is `from_path` and `from_filename`. `from_filename` picks up the first matching filename, which may be handy if you are running `cargo` from different places and want to use relative paths to your *.env* files.

Hope this helps! Let me know if you have any other questions.

**Reply by SquireOfSoftware** [reply]
Ah cool, thanks for all the answers and the link to the `docs.rs` site!
